engine/common/synchronization/engine_spam_test.go

```diff
@@ -0,0 +1,259 @@
+			for i := 0; i < load; i++ {
+				ss.T().Log("load iteration", i)
+				nonce, err := rand.Uint64()
+				require.NoError(ss.T(), err, "should generate nonce")
+
+				// generate origin and request message
+				originID := unittest.IdentifierFixture()
+				req := &messages.SyncRequest{
+					Nonce:  nonce,
+					Height: 0,
+				}
+
+				// if request height is higher than local finalized, we should not respond
+				req.Height = ss.head.Height + 1
+
+				ss.core.On("HandleHeight", ss.head, req.Height)
+				ss.core.On("WithinTolerance", ss.head, req.Height).Return(false)
+				ss.con.AssertNotCalled(ss.T(), "Unicast", mock.Anything, mock.Anything)
+
+				// maybe function calls that might or might not occur over the course of the load test
+				ss.core.On("ScanPending", ss.head).Return([]chainsync.Range{}, []chainsync.Batch{}).Maybe()
+				ss.con.On("Multicast", mock.Anything, mock.Anything, mock.Anything, mock.Anything).Return(nil).Maybe()
+
+				// count misbehavior reports over the course of a load test
+				ss.con.On("ReportMisbehavior", mock.Anything).Return(mock.Anything).Maybe().Run(
+					func(args mock.Arguments) {
+						misbehaviorsCounter++
+					},
+				)
+				ss.e.spamDetectionConfig.syncRequestProbability = loadGroup.syncRequestProbabilityFactor
+				require.NoError(ss.T(), ss.e.Process(channels.SyncCommittee, originID, req))
+			}
+
+			// check function call expectations at the end of the load test; otherwise, load test would take much longer
+			ss.core.AssertExpectations(ss.T())
+			ss.con.AssertExpectations(ss.T())
+
+			// check that correct range of misbehavior reports were generated
+			// since we're using a probabilistic approach to generate misbehavior reports, we can't guarantee the exact number,
+			// so we check that it's within an expected range
+			ss.T().Logf("misbehaviors counter after load test: %d (expected lower bound: %d expected upper bound: %d)", misbehaviorsCounter, loadGroup.expectedMisbehaviorsLower, loadGroup.expectedMisbehaviorsUpper)
+			assert.GreaterOrEqual(ss.T(), misbehaviorsCounter, loadGroup.expectedMisbehaviorsLower)
+			assert.LessOrEqual(ss.T(), misbehaviorsCounter, loadGroup.expectedMisbehaviorsUpper)
+
+			misbehaviorsCounter = 0 // reset counter for next subtest
+		})
+	}
+}
+
+// TestLoad_Process_RangeRequest_SometimesReportSpam is a load test that ensures that a misbehavior report is generated
+// an appropriate range of times when the base probability factor and range are set to different values.
+func (ss *SyncSuite) TestLoad_Process_RangeRequest_SometimesReportSpam() {
+	ctx, cancel := irrecoverable.NewMockSignalerContextWithCancel(ss.T(), context.Background())
+	ss.e.Start(ctx)
+	unittest.AssertClosesBefore(ss.T(), ss.e.Ready(), time.Second)
+	defer cancel()
+
+	load := 1000
+
+	type loadGroup struct {
+		rangeRequestBaseProb      float32
+		expectedMisbehaviorsLower int
+		expectedMisbehaviorsUpper int
+		fromHeight                uint64
+		toHeight                  uint64
+	}
```

Comment:
For this test and other tests above, please explain what each of these attributes means.

Reply:
added godoc to explain loadGroup fields and how load tests work

3920b4d3ca86ad71fd531fb1b704ad7023f60a02